File: Algorithms/1048_Longest_String_Chain/Python/Longest_String_Chain_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def longestStrChain(self, words):
        length = len(words)
        if length == 1: return 1
        cache = {}
        words = sorted(words, key=lambda x: len(x))
        res = 1

        for word in words:
            cache[word] = 1
            word_length = len(word)
            if word_length == 1: continue
            for i in range(word_length):
                temp_word = word[:i] + word[i + 1:]
                if temp_word in cache:
                    cache[word] = max(cache[word], cache[temp_word] + 1)
                    res = max(res,cache[word])

        return res

File: Algorithms/1048_Longest_String_Chain/Java/LongestStringChainSolution1.java

// Space: O(n)
// Time: O(n)

public class Solution {

    public int longestStrChain(String[] words) {
        if (words.length == 1) return 1;
        int res = 0;
        Arrays.sort(words, (s1, s2) -> {
            return s1.length() - s2.length();
        });
        Map<String, Integer> cache = new HashMap<>();

        for (String word : words) {
            cache.put(word, 1);
            if (word.length() == 1) continue;

            for (int i = 0; i < word.length(); i++) {
                String temp = word.substring(0, i) + word.substring(i + 1);
                if (cache.containsKey(temp)) {
                    cache.put(word, Math.max(cache.get(word), cache.get(temp) + 1));
                    res = Math.max(cache.get(word), res);
                }
            }
        }
        return res;
    }
}




